### PR TITLE
build: Add CMake-based build system (5 of N) 

### PR DESCRIPTION
**Pull request by hebasto**
It was [suggested](https://github.com/hebasto/bitcoin/pull/10#issuecomment-1492012002):

> It might be helpful to get some more of the depends interactions hooked up as a next step

Done in this PR.

---

The parent PR: https://github.com/bitcoin/bitcoin/pull/25797.
The previous PRs in the staging branch: https://github.com/hebasto/bitcoin/pull/5, https://github.com/hebasto/bitcoin/pull/6, https://github.com/hebasto/bitcoin/pull/7, https://github.com/hebasto/bitcoin/pull/10.

---

EXAMPLES:

Cross-compiling for Windows:
```
make -C depends HOST=x86_64-w64-mingw32 NO_QT=1 NO_WALLET=1
cmake -S . -B build -DCMAKE_TOOLCHAIN_FILE=depends/x86_64-w64-mingw32/share/toolchain.cmake
cmake --build build
```

Cross-compiling for macOS, arm64:
```
make -C depends HOST=arm64-apple-darwin NO_QT=1 NO_WALLET=1
cmake -S . -B build -DCMAKE_TOOLCHAIN_FILE=depends/arm64-apple-darwin/share/toolchain.cmake
cmake --build build
```

---

Drafted as currently it is based on https://github.com/hebasto/bitcoin/pull/11.

### CODE REVIEW

File: cmake/optional.cmake [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# Optional features and packages.
+
+if(CCACHE)
+  find_program(CCACHE_EXECUTABLE ccache)
+  if(CCACHE_EXECUTABLE)
+    set(CCACHE ON)
+    if(MSVC)
+      # See https://github.com/ccache/ccache/wiki/MS-Visual-Studio
+      set(MSVC_CCACHE_WRAPPER_CONTENT "\"${CCACHE_EXECUTABLE}\" \"${CMAKE_CXX_COMPILER}\"")
+      set(MSVC_CCACHE_WRAPPER_FILENAME wrapped-cl.bat)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/optional.cmake:14]: It is a replacement for our current `ci/test/wrapped-cl.bat`.

File: cmake/module/FindNATPMP.cmake [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  NAMES natpmp
+  HINTS ${natpmp_brew_prefix}/lib
+)
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(NATPMP
+  REQUIRED_VARS NATPMP_LIBRARY NATPMP_INCLUDE_DIR
+)
+
+if(NATPMP_FOUND AND NOT TARGET NATPMP::NATPMP)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/module/FindNATPMP.cmake:29]: > What does this condition represent?

We consider the package `NATPMP` as found and check that the `NATPMP::NATPMP` was not created previously to do not conflict in the next line.

File: CMakeLists.txt [hunk 3]
```diff
@@ -44,6 +44,7 @@ cmake_dependent_option(CXX20 "Enable compilation in C++20 mode." OFF "NOT MSVC"
 option(THREADLOCAL "Enable features that depend on the C++ thread_local keyword (currently just thread names in debug logs)." ON)
 include(TristateOption)
 tristate_option(CCACHE "Use ccache for compiling." "if ccache is found." AUTO)
+tristate_option(WITH_NATPMP "Enable NAT-PMP." "if libnatpmp is found." AUTO)
```
[the remark this answers is not in the document]
Comment (fanquake) [reply CMakeLists.txt:47]: Yea. We certainly wont be adopting this `tri-state` logic/machinery into our CMake build system long-term/at-merge-time. I'd consider it a regression to reimplement this AutoTools behaviour. As for disabling certain "features" by default, yes, I thin that is what we are going to do. It continues to make less & less sense to compile in stuff like this by default, let alone auto-opt-in builders to it, based on what happens to be installed on their systems at the time of compilation.

**Comment by hebasto** [comment]
Moved to #13 after rebasing from the master branch to new cmake-staging one.

File: cmake/module/FindNATPMP.cmake [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2023 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+if(CMAKE_HOST_APPLE)
+  execute_process(
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/module/FindNATPMP.cmake]: Done in https://github.com/hebasto/bitcoin/pull/13.

File: cmake/module/FindMiniUPnPc.cmake [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+endif()
+
+find_path(MiniUPnPc_INCLUDE_DIR
+  NAMES miniupnpc/miniupnpc.h
+  PATHS ${PC_MiniUPnPc_INCLUDE_DIRS}
+  HINTS ${miniupnp_brew_prefix}/include
+)
+
+if(MiniUPnPc_INCLUDE_DIR)
+  file(
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/module/FindMiniUPnPc.cmake:20]: Going to address this comment in #13 (https://github.com/hebasto/bitcoin/pull/13#issuecomment-1498260574).

**Pull request closed without merging**